src/r1cs/constraint_system.rs

```diff
@@ -1,8 +1,56 @@
 //! Definition of the constraint system trait.
 
+use std::collections::HashSet;
+
 use super::{LinearCombination, R1CSError, Variable};
 use curve25519_dalek::scalar::Scalar;
 use merlin::Transcript;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug, Default)]
+pub struct SparseWeightRow(pub Vec<(usize, Scalar)>);
+
+impl PartialEq for SparseWeightRow {
+    fn eq(&self, other: &Self) -> bool {
+        // Because sparse constraint representation is constructed by iterating over (K, V) pairs
+        // of underlying linear combinations, order of terms in the matrix is not guaranteed.
+        // This is okay, because the matrix elements contain the index of the variable
+        // to which the given weight is applied, but requires us to have an
+        // ordering-agnostic equality relation
+        let a: HashSet<&(usize, Scalar)> = self.0.iter().collect();
+        let b: HashSet<&(usize, Scalar)> = other.0.iter().collect();
+
+        a == b
+    }
+}
+impl Eq for SparseWeightRow {}
+
+impl Extend<(usize, Option<Scalar>)> for SparseWeightRow {
+    fn extend<T: IntoIterator<Item = (usize, Option<Scalar>)>>(&mut self, iter: T) {
```

Comment:
Why is this an `Option`? Can we add a comment? Or better wrap this in a named struct `SparseRowEntry` to make the code more readable

Reply:
added a comment - this is more to handle building the `c` vector so it wouldn't make sense to call `(usize, Option<Scalar>)` a `SparseRowEntry`, that should just be a `(usize, Scalar)`